src/internal/kopia/snapshot_manager.go

```diff
@@ -0,0 +1,232 @@
+		m := foundMans[metas[i].ID]
+		if m == nil || len(m.IncompleteReason) > 0 {
+			continue
+		}
+
+		return i
+	}
+
+	return -1
+}
+
+// findRecentManifests searches through mans and returns the most recent
+// complete manifest (if one exists) and maybe the most recent incomplete
+// manifest. If the newest incomplete manifest is more recent than the newest
+// complete manifest then adds it to the returned list. Otherwise no incomplete
+// manifest is returned. Returns nil if there are no complete or complete
+// manifests in mans.
+func findRecentManifests(
+	mans []*snapshot.Manifest,
+) []*snapshot.Manifest {
+	var (
+		res             []*snapshot.Manifest
+		foundIncomplete = false
+	)
+
+	// Manifests should maintain the sort order of the original IDs that were used
+	// to fetch the data, but just in case sort oldest to newest.
+	mans = snapshot.SortByTime(mans, false)
+
+	for i := len(mans) - 1; i >= 0; i-- {
+		m := mans[i]
+
+		if len(m.IncompleteReason) > 0 {
+			if !foundIncomplete {
+				foundIncomplete = true
+
+				res = append(res, m)
+			}
+
+			continue
+		}
+
+		// Once we find a complete snapshot we're done, even if we haven't
+		// found an incomplete one yet.
+		res = append(res, m)
+
+		break
+	}
+
+	return res
+}
+
+// fetchPrevManifests returns the most recent, as-of-yet unfound complete and
+// (maybe) incomplete manifests in metas. If the most recent incomplete manifest
+// is older than the most recent complete manifest no incomplete manifest is
+// returned. If only incomplete manifests exists, returns the most recent one.
+// Returns no manifests if an error occurs.
+func fetchPrevManifests(
+	ctx context.Context,
+	sm snapshotManager,
+	foundMans map[manifest.ID]*snapshot.Manifest,
+	tags map[string]string,
+) ([]*snapshot.Manifest, error) {
+	metas, err := sm.FindManifests(ctx, tags)
+	if err != nil {
+		return nil, errors.Wrap(err, "fetching manifest metas by tag")
+	}
+
+	if len(metas) == 0 {
+		return nil, nil
+	}
+
+	lastCompleteIdx := getLastIdx(foundMans, metas)
+
+	// We have a complete cached snapshot and it's the most recent. No need
+	// to do anything else.
+	if lastCompleteIdx == len(metas)-1 {
+		return nil, nil
+	}
+
+	// TODO(ashmrtn): Remainder of the function can be simplified if we can inject
+	// different tags to the snapshot checkpoints than the complete snapshot.
+
+	// Fetch all manifests newer than the oldest complete snapshot. A little
+	// wasteful as we may also re-fetch the most recent incomplete manifest, but
+	// it reduces the complexity of returning the most recent incomplete manifest
+	// if it is newer than the most recent complete manifest.
+	ids := make([]manifest.ID, 0, len(metas)-(lastCompleteIdx+1))
+	for i := lastCompleteIdx + 1; i < len(metas); i++ {
+		ids = append(ids, metas[i].ID)
+	}
+
+	mans, err := sm.LoadSnapshots(ctx, ids)
+	if err != nil {
+		return nil, errors.Wrap(err, "fetching previous manifests")
+	}
+
+	return findRecentManifests(mans), nil
```

Comment:
on line 160, you return `nil` if the most recent snapshot is complete; ie, not returning a manifest.  Here, you're returning (potentially) the most recent complete manifest plus the most recent incomplete one.  Feels like a bit of a discrepancy.  What is the reason to not return the recent completed manifest on 160?

Reply:
ah, so there's two parts to this function:
1. see if we've already got a cached version of the manifest in `foundMans`
2. fetch newer manifests if we don't have the most recent manifest cached

Line 160 handles the case where we had a cached manifest _and_ it was the most recent. If we don't hit that then this will fetch all manifests newer than the cached one we found and try to find a complete manifest and/or an incomplete manifest

Reply:
> we had a cached manifest and it was the most recent

So in this case the caller is expected to already know, of the cached manifests that it has, which is the most recent?

Reply:
mmm I suppose? I would view this more as a helper function that's only going to be called by a single caller (the one that's calling it now). All it needs to do is return any currently uncached manifests to the caller. The caller will add the returned manifests to its set and eventually those are all passed to kopia which doesn't really care if it's only the most recent manifests or not

Given that the manifests are being added to a map anyway, I could return the cached manifest on line 160, but it doesn't really matter one way or the other in the long run

Reply:
okedoke, that makes sense